packages/doenetml-worker-rust/doenetml-core/src/state_var_calculations.rs

```diff
@@ -0,0 +1,762 @@
+                                    // Recurse by putting two items on the stack
+                                    // and then continuing at start of stack loop
+                                    freshen_stack.push(StateVarCalculationState::Stale(
+                                        StaleCalculationState {
+                                            component_state,
+                                            instruction_idx,
+                                            val_idx,
+                                        },
+                                    ));
+                                    freshen_stack.push(StateVarCalculationState::Stale(
+                                        StaleCalculationState {
+                                            component_state: new_component_state,
+                                            instruction_idx: 0,
+                                            val_idx: 0,
+                                        },
+                                    ));
+
+                                    // break out of the inner loop and continue to outer loop,
+                                    // which will result in first popping off the last instruction
+                                    // that we added to the stack
+                                    continue 'stack_loop;
+                                }
+                                Freshness::Unresolved => {
+                                    panic!("How did a stale state variable depend on an unresolved state variable?")
+                                }
+                            };
+                        }
+                    }
+
+                    initial_val_idx = 0;
+                }
+
+                // If we have gotten to this part,
+                // then we have freshened all dependencies
+                // for the state variable (component_idx, state_var_idx).
+                // We calculate its value if a dependency has changed,
+                // else restore its previous value.
+
+                let component_idx = component_state.component_idx;
+                let state_var_idx = component_state.state_var_idx;
+
+                let mut comp = components[component_idx].borrow_mut();
+                let state_var = &mut comp.get_state_variables_mut()[state_var_idx];
+
+                // Check if any dependency has changed since we last called record_all_dependencies_viewed.
+                if state_var.check_if_any_dependency_changed_since_last_viewed() {
+                    state_var.calculate_state_var_from_dependencies();
+                    state_var.record_all_dependencies_viewed();
+                } else {
+                    state_var.restore_previous_value();
+                }
+            }
+        }
+    }
+}
+
+/// Freshen the state variable specified by original_component_state,
+/// then get its fresh value
+pub fn get_state_var_value(
```

Comment:
Since this file is long, I think external functions should be moved to their own file.

Reply:
Separated the state variable calculates and the state variable updates into two files.